README.md
# ubuntukylin-theme

　　ubuntukylin-theme 优麒麟主题包，此包包含开机动画、gtk主题、图标主题、光标主题、开机选择界面和搜狗输入法皮肤。

一、开机动画

　　控制开机动画文件在usr/share/plymouth/下，首先如何查看当前系统开机动画呢？重启不是一个好办法，你可以进入root权限后在终端输入:

	plymouthd

　　然后切到tty1-6任意，再次进入root权限并输入：

	plymouth --show-splash

　　就可以看见当前开机动画了，如果你想退出预览直接切回tty7即可。

　　如果你需要替换开机动画，那么在usr/share/plymouth/themes/ubuntukylin-logo/下存放着开机动画每帧的图片，连起来就是动画了，控制图片连起来的就是ubuntukylin-logo.script，如果图片数量发生改变请在此文件下对应值进行修改。当你修改完成后请放在系统对应目录下，然后再次执行命令查看开机动画。

二、gtk主题

　　想要修改gtk主题首先需要给系统安装sass，直接在终端输入命令：

	sudo apt install sass

　　然后建议学习一下sass，当然如果你熟悉css可以很快掌握，可以直接查看官网的教程(https://www.sass.hk/guide/)，也可以去看看阮一峰的教程(http://www.ruanyifeng.com/blog/2012/06/sass.html)。

　　学过CSS的人都知道，它不是一种编程语言。你可以用它开发网页样式，但是没法用它编程。也就是说，CSS基本上是设计师的工具，不是程序员的工具。所以这里使用的是sass，目的就是为了更加配合程序猿的使用习惯，使代码更有逻辑性。当然为了使系统能够读取，必须得在编包安装时把.scss文件转换成css。

　　在你修改主题之前要清楚一点，所谓gtk主题的错误并不会导致系统崩溃，但是可能会因为颜色的配置给人一种“某个按钮无法生效、某个控件没有被激活”这样的感受，这也是在修改时要注意的，把每个控件的各种状态情况考虑，如果某个控件完全没有样式或者样式完全没有被控制。

　　当某个控件的样式出现问题，可以通过删除代码或者修改颜色色值来定位，注意sass存在继承关系，如果要修改的控件被其他控件继承，会导致很大的影响。当然如果单独给某个控件颜色时，请把颜色写在对应控制颜色的文件下，这样会方便后续的维护。

　　当你修改完毕，想要尝试修改是否生效，可以直接在含有Makefile文件的文件下执行：

	make

	sudo make install

　　即可，但要注意的是在往github提交代码时请执行：

	make clean

　　清除make产生的多余文件，这对编包上传等工作百利无一害。


三、图标主题

　　首先在修改图标主题中的图片时，请遵守对图标命名的规范，最基础的就是不向文件夹里乱添加图片。为了使用户再切换主题时每个图标都能生效，这是必须的。

　　当你打开图片文件夹时，你可以看见很多按照规格排列的文件夹。是的，这就是图片大小，当你得到一张.png切符合命名规范的图片想要
替换到图标主题时，也请根据尺寸到相应文件夹以及相应类型下进行替换。请注意，scalable文件下没有尺寸，因为此文件夹下的图片为.svg
类型，.svg类型图片没有固定大小，颜色也可以根据程序进行改变。

　　目录下的index.theme文件是配置文件。如果你检查到明明对应文件夹里有图片却一直没有显示或者显示尺寸有错，那么可能是这里出错了，每个目录下的文件夹都会被此文件配置，如果有新增加的文件夹那么请一定要更新此配置文件，不然此文件夹下的所有内容都不会被调用。

　　那么到底要如何替换图标呢？截止目前为止，此目录下有170+的文件夹，有超过2.7万张图片，寻找某一张图片的无疑是大海捞针。方法一，在/usr/share/application目录下有对应的.desktop文件，打开对应软件的.desktop文件，可以看见Icon一行，后面接的就是对应软件名字；方法二，清楚需要替换的图片是属于什么类型，直接在对应类型的文件夹下寻找；这两种方法基本可以帮助你定位出错图片位置，当无法确认对应文件类型时就用排除法，删掉文件夹，如果被影响了，就说明图片在此文件夹下。

　　在图标主题下有很多相同的图标，那么我们该如何控制他们同时切换呢，这个时候就用到了软链接，因软链接生成的图片会跟着父级图片进行改变。为了方便批量替换，我写了一个脚本<https://github.com/kwwzzz/shell/tree/master/ukui-theme-icons-replace>用来替换同名的图标，这样能够节约部分时间提高效率，毕竟总是重复的工作会让人感觉枯燥。

　　关于图标命名规范，首先是和上游的图标名称相同，还有图标名中不可以出现空格，不可以出现奇怪的符号，图片的权限也必须合理等等，你可以通过编包，跑lintian来检查错误，由于主题包过大，编包跑lintian往往要很久时间，所以我写了一个脚本用于检查修改图片权限<https://github.com/kwwzzz/shell/tree/master/ukui-theme-icons-power>，这样也能够省去一些重复的编包测试工作。

四、光标主题

　　关于光标主题，为了让ukui的鼠标更有个性化，我调研了ubuntu的光标主题并以ubuntu为基础进行ukui的光标主题制作。

　　光标主题读取的不是简单的.png图片，是经过配置后的鼠标文件，直接使用图像工具无法打开，目前使用了gimp可以看见鼠标文件。并且尝试是否可以直接用gimp绘制鼠标文件，目前是不可行的，目前只有一种办法制作鼠标文件。

　　首先获取到所有光标名称，ubuntu下是90个，但由于软链接其实只有50种需要修改，现已经提取出这些鼠标文件名称（left_ptr是默认光标），使用gimp可以打开这些鼠标文件，绘制出对应合适的外观。该鼠标文件分为静态和动态，静态是由一张图片的三种尺寸组成（48x48px,32x32px,24x24px）共三张即可,动态同样是按照三种尺寸组成，但还需要按照帧数进行排列（ubuntu是30张图片完成一个动画），目前只有两张图片是动态（watch、left_ptr_watch），注意静态图片命名请按照：图片名_尺寸.png(如：left_ptr_48.png)，动态图片按照：图片名_序号_尺寸.png(如：watch_1_48.png、watch_2_48.png)。

　　绘制完图片后，制作.conf文件,內容格式為：

　　(size) (xhot) (yhot) (filename) (ms-delay)

　　其中：

　　(size)处写鼠标的大小（像素）

　　(xhot)(yhot)鼠标的有效区域以及偏移量

　　(filename)图片名

　　(ms-delay)动态鼠标使用，每帧图片的间隔时间

　　制作.conf后，通过命令制作成系统可读取的文件。生成命令：

　　xcursorgen xxxx.conf xxxx

　　以上是光标制作的具体流程，为了更加直观以及方便修改光标主题，我已经将光标制作自动化，如果有需要变化图片的光标直接在对应尺寸下替换图片即可，通过

	make

	make install

　　命令生成对应的光标主题，注意提交前执行

	make clean

　　删除多余文件


debian/changelog
ubuntukylin-theme (3.0.0~0802) focal; urgency=medium

  * New release.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Mon, 02 Aug 2021 14:14:53 +0800

ubuntukylin-theme (2.0.5) focal; urgency=medium

  * Bugfix only:
    - Fix the problem that the webpage prompt is fuzzy and unclear.
    - Fix the error in the checkbox and raidobox.
    - Fix the disorder of color structure.
    - Fix the background overlap problem of the drop-down menu.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Mon, 06 Apr 2020 20:06:49 +0800

ubuntukylin-theme (2.0.4) focal; urgency=medium

  * Bugfix only:
    - Fix that input box style missing.
    - Fix menu color being covered.
    - Fix window background shadow lose.
    - Fix scroll bar missing color.
    - Fix a syntax error.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Mon, 30 Mar 2020 22:31:07 +0800

ubuntukylin-theme (2.0.3) focal; urgency=medium

  * Update the boot animation.
  * Update the grub theme.
  * Update the ukui-white theme.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Mon, 23 Mar 2020 23:09:05 +0800

ubuntukylin-theme (2.0.2) focal; urgency=medium

  * Update the gtk style and update some icons.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Tue, 17 Mar 2020 10:40:50 +0800

ubuntukylin-theme (2.0.1.1) focal; urgency=medium

  * debian/rules:
    - Readd missing NO_PNG_PKG_MANAGLE=1.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Wed, 26 Feb 2020 14:02:22 +0800

ubuntukylin-theme (2.0.1) focal; urgency=medium

  * Remove the empty images.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Wed, 26 Feb 2020 10:42:27 +0800

ubuntukylin-theme (2.0.0) focal; urgency=medium

  * Update style and icons for ukui3.0.
  * debian:
    - Bump standards-vertion to 4.5.0.
    - Use debhelper-compat.
    - Set Rules-Requires-Root to no.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Tue, 25 Feb 2020 23:34:47 +0800

ubuntukylin-theme (1.8.3) eoan; urgency=medium

  * Add missing icons of media-cdr, image magic and software center.
  * Update some icons of panel.
  * Fix wrong mode of index.theme.in.
  * Debian/control:
    - Bump standards-vertion to 4.4.1.

 -- handsome_feng <jianfengli@ubuntukylin.com>  Mon, 07 Oct 2019 12:02:59 +0800


ubuntukylin-theme (1.8.2.1) disco; urgency=medium

  * Add generic-x-text, generic-x-script, folder-drag-accept,
    blueman-active and nm-device-wired, drive-hardist,
    timing-shutdown icons.

ukui-blue/gtk-3.0/css-optimizer.js
'use strict';

var fs = require('fs');
try {
	var css_data = fs.readFileSync('gtk-contained.css', 'utf-8');
} catch (err) {
	console.log("Read css file error.");
}

var CleanCSS = require('clean-css');
var input = 'a{font-weight:bold;}'; /* for test */
var options = {
	format: {
		breaks: {
			afterAtRule: true,
			afterBlockBegins: true,
			afterBlockEnds: true,
			afterComment: true,
			afterProperty: true,
			afterRuleBegins: true,
			afterRuleEnds: true,
			beforeBlockEnds: true,
			betweenSelectors: true
		},
		indentBy: 4,
		indentWith: 'space',
		spaces: {
			aroundSelectorRelation: true,
			beforeBlockBegins: true,
			beforeValue: true
		},
		wrapAt: 80
	},
	level: {
		1: {
			all: false,
			specialComments: 'all'
		}
	}
};

var output = new CleanCSS(options).minify(css_data);
//console.log(output.styles);
fs.writeFileSync("gtk-contained.css", output.styles);


dark-sense/cursors.sh
#!/bin/bash
  mkdir cursors
  echo "24 11 11 ../24x24/bd_double_arrow.png 0 "> cursors/bd_double_arrow.in
  echo "24 4 20 ../24x24/bottom_left_corner.png 0 "> cursors/bottom_left_corner.in
  echo "24 20 20 ../24x24/bottom_right_corner.png 0 "> cursors/bottom_right_corner.in
  echo "24 11 18 ../24x24/bottom_side.png 0 "> cursors/bottom_side.in
  echo "24 11 18 ../24x24/bottom_tee.png 0 "> cursors/bottom_tee.in
  echo "24 4 5 ../24x24/circle.png 0 "> cursors/circle.in
  echo "24 4 19 ../24x24/color-picker.png 0 "> cursors/color-picker.in
  echo "24 4 5 ../24x24/copy.png 0 "> cursors/copy.in
  echo "24 11 11 ../24x24/cross.png 0 "> cursors/cross.in
  echo "24 11 11 ../24x24/crossed_circle.png 0 "> cursors/crossed_circle.in
  echo "24 11 11 ../24x24/crosshair.png 0 "> cursors/crosshair.in
  echo "24 9 9 ../24x24/dnd-ask.png 0 "> cursors/dnd-ask.in
  echo "24 9 9 ../24x24/dnd-copy.png 0 "> cursors/dnd-copy.in
  echo "24 9 9 ../24x24/dnd-link.png 0 "> cursors/dnd-link.in
  echo "24 9 9 ../24x24/dnd-move.png 0 "> cursors/dnd-move.in
  echo "24 13 11 ../24x24/dnd-none.png 0 "> cursors/dnd-none.in
  echo "24 11 11 ../24x24/dotbox.png 0 "> cursors/dotbox.in
  echo "24 11 11 ../24x24/fd_double_arrow.png 0 "> cursors/fd_double_arrow.in
  echo "24 13 11 ../24x24/grabbing.png 0 "> cursors/grabbing.in
  echo "24 9 5 ../24x24/hand2.png 0 "> cursors/hand2.in
  echo "24 7 4 ../24x24/left_ptr.png 0 "> cursors/left_ptr.in
  echo "24 7 13 ../24x24/left_side.png 0 "> cursors/left_side.in
  echo "24 7 13 ../24x24/left_tee.png 0 "> cursors/left_tee.in
  echo "24 4 5 ../24x24/link.png 0 "> cursors/link.in
  echo "24 4 20 ../24x24/ll_angle.png 0 "> cursors/ll_angle.in
  echo "24 20 20 ../24x24/lr_angle.png 0 "> cursors/lr_angle.in
  echo "24 4 5 ../24x24/move.png 0 "> cursors/move.in
  echo "24 7 23 ../24x24/pencil.png 0 "> cursors/pencil.in
  echo "24 11 11 ../24x24/plus.png 0 "> cursors/plus.in
  echo "24 12 21 ../24x24/question_arrow.png 0 "> cursors/question_arrow.in
  echo "24 17 4 ../24x24/right_ptr.png 0 "> cursors/right_ptr.in
  echo "24 20 13 ../24x24/right_side.png 0 "> cursors/right_side.in
  echo "24 20 13 ../24x24/right_tee.png 0 "> cursors/right_tee.in
  echo "24 11 18 ../24x24/sb_down_arrow.png 0 "> cursors/sb_down_arrow.in
  echo "24 11 11 ../24x24/sb_h_double_arrow.png 0 "> cursors/sb_h_double_arrow.in
  echo "24 7 13 ../24x24/sb_left_arrow.png 0 "> cursors/sb_left_arrow.in
  echo "24 20 13 ../24x24/sb_right_arrow.png 0 "> cursors/sb_right_arrow.in
  echo "24 12 4 ../24x24/sb_up_arrow.png 0 "> cursors/sb_up_arrow.in
  echo "24 11 11 ../24x24/sb_v_double_arrow.png 0 "> cursors/sb_v_double_arrow.in
  echo "24 11 11 ../24x24/tcross.png 0 "> cursors/tcross.in
  echo "24 5 5 ../24x24/top_left_corner.png 0 "> cursors/top_left_corner.in
  echo "24 18 5 ../24x24/top_right_corner.png 0 "> cursors/top_right_corner.in
  echo "24 12 5 ../24x24/top_side.png 0 "> cursors/top_side.in
  echo "24 12 5 ../24x24/top_tee.png 0 "> cursors/top_tee.in
  echo "24 5 5 ../24x24/ul_angle.png 0 "> cursors/ul_angle.in
  echo "24 18 5 ../24x24/ur_angle.png 0 "> cursors/ur_angle.in
  echo "24 11 11 ../24x24/X_cursor.png 0 "> cursors/X_cursor.in
  echo "24 11 11 ../24x24/xterm.png 0 "> cursors/xterm.in
  echo "24 11 11 ../24x24/zoom-in.png 0 "> cursors/zoom-in.in
  echo "24 11 11 ../24x24/zoom-out.png 0 "> cursors/zoom-out.in

  echo "32 15 15 ../32x32/bd_double_arrow.png 0 ">> cursors/bd_double_arrow.in
  echo "32 6 30 ../32x32/bottom_left_corner.png 0 ">> cursors/bottom_left_corner.in
  echo "32 28 28 ../32x32/bottom_right_corner.png 0 ">> cursors/bottom_right_corner.in
  echo "32 15 20 ../32x32/bottom_side.png 0 ">> cursors/bottom_side.in
  echo "32 15 20 ../32x32/bottom_tee.png 0 ">> cursors/bottom_tee.in
  echo "32 6 8 ../32x32/circle.png 0 ">> cursors/circle.in
  echo "32 6 26 ../32x32/color-picker.png 0 ">> cursors/color-picker.in
  echo "32 6 8 ../32x32/copy.png 0 ">> cursors/copy.in
  echo "32 15 15 ../32x32/cross.png 0 ">> cursors/cross.in
  echo "32 15 15 ../32x32/crossed_circle.png 0 ">> cursors/crossed_circle.in
  echo "32 15 15 ../32x32/crosshair.png 0 ">> cursors/crosshair.in
  echo "32 13 13 ../32x32/dnd-ask.png 0 ">> cursors/dnd-ask.in
  echo "32 13 13 ../32x32/dnd-copy.png 0 ">> cursors/dnd-copy.in
  echo "32 13 13 ../32x32/dnd-link.png 0 ">> cursors/dnd-link.in
  echo "32 13 13 ../32x32/dnd-move.png 0 ">> cursors/dnd-move.in
  echo "32 20 17 ../32x32/dnd-none.png 0 ">> cursors/dnd-none.in
  echo "32 15 15 ../32x32/dotbox.png 0 ">> cursors/dotbox.in
  echo "32 15 15 ../32x32/fd_double_arrow.png 0 ">> cursors/fd_double_arrow.in
  echo "32 20 17 ../32x32/grabbing.png 0 ">> cursors/grabbing.in
  echo "32 14 8 ../32x32/hand2.png 0 ">> cursors/hand2.in
  echo "32 10 5 ../32x32/left_ptr.png 0 ">> cursors/left_ptr.in
  echo "32 10 20 ../32x32/left_side.png 0 ">> cursors/left_side.in
  echo "32 10 20 ../32x32/left_tee.png 0 ">> cursors/left_tee.in
  echo "32 6 8 ../32x32/link.png 0 ">> cursors/link.in
  echo "32 6 30 ../32x32/ll_angle.png 0 ">> cursors/ll_angle.in
  echo "32 28 28 ../32x32/lr_angle.png 0 ">> cursors/lr_angle.in
  echo "32 6 8 ../32x32/move.png 0 ">> cursors/move.in
  echo "32 9 31 ../32x32/pencil.png 0 ">> cursors/pencil.in
  echo "32 15 15 ../32x32/plus.png 0 ">> cursors/plus.in
  echo "32 16 30 ../32x32/question_arrow.png 0 ">> cursors/question_arrow.in
  echo "32 24 6 ../32x32/right_ptr.png 0 ">> cursors/right_ptr.in
  echo "32 30 19 ../32x32/right_side.png 0 ">> cursors/right_side.in
  echo "32 30 19 ../32x32/right_tee.png 0 ">> cursors/right_tee.in
  echo "32 15 20 ../32x32/sb_down_arrow.png 0 ">> cursors/sb_down_arrow.in
  echo "32 15 15 ../32x32/sb_h_double_arrow.png 0 ">> cursors/sb_h_double_arrow.in
  echo "32 10 20 ../32x32/sb_left_arrow.png 0 ">> cursors/sb_left_arrow.in
  echo "32 30 19 ../32x32/sb_right_arrow.png 0 ">> cursors/sb_right_arrow.in
  echo "32 18 6 ../32x32/sb_up_arrow.png 0 ">> cursors/sb_up_arrow.in
  echo "32 15 15 ../32x32/sb_v_double_arrow.png 0 ">> cursors/sb_v_double_arrow.in
  echo "32 15 15 ../32x32/tcross.png 0 ">> cursors/tcross.in
  echo "32 8 8 ../32x32/top_left_corner.png 0 ">> cursors/top_left_corner.in
  echo "32 26 8 ../32x32/top_right_corner.png 0 ">> cursors/top_right_corner.in
  echo "32 18 8 ../32x32/top_side.png 0 ">> cursors/top_side.in
  echo "32 18 8 ../32x32/top_tee.png 0 ">> cursors/top_tee.in
  echo "32 8 8 ../32x32/ul_angle.png 0 ">> cursors/ul_angle.in
  echo "32 26 8 ../32x32/ur_angle.png 0 ">> cursors/ur_angle.in
  echo "32 15 15 ../32x32/X_cursor.png 0 ">> cursors/X_cursor.in
  echo "32 15 15 ../32x32/xterm.png 0 ">> cursors/xterm.in
  echo "32 15 15 ../32x32/zoom-in.png 0 ">> cursors/zoom-in.in
  echo "32 15 15 ../32x32/zoom-out.png 0 ">> cursors/zoom-out.in

  echo "48 22 22 ../48x48/bd_double_arrow.png 0 ">> cursors/bd_double_arrow.in
  echo "48 8 40 ../48x48/bottom_left_corner.png 0 ">> cursors/bottom_left_corner.in
  echo "48 40 40 ../48x48/bottom_right_corner.png 0 ">> cursors/bottom_right_corner.in
  echo "48 22 36 ../48x48/bottom_side.png 0 ">> cursors/bottom_side.in
  echo "48 22 36 ../48x48/bottom_tee.png 0 ">> cursors/bottom_tee.in
  echo "48 8 10 ../48x48/circle.png 0 ">> cursors/circle.in
  echo "48 10 38 ../48x48/color-picker.png 0 ">> cursors/color-picker.in
  echo "48 8 10 ../48x48/copy.png 0 ">> cursors/copy.in
  echo "48 22 22 ../48x48/cross.png 0 ">> cursors/cross.in
  echo "48 22 22 ../48x48/crossed_circle.png 0 ">> cursors/crossed_circle.in
  echo "48 22 22 ../48x48/crosshair.png 0 ">> cursors/crosshair.in
  echo "48 18 18 ../48x48/dnd-ask.png 0 ">> cursors/dnd-ask.in
  echo "48 18 18 ../48x48/dnd-copy.png 0 ">> cursors/dnd-copy.in
  echo "48 18 18 ../48x48/dnd-link.png 0 ">> cursors/dnd-link.in
  echo "48 18 18 ../48x48/dnd-move.png 0 ">> cursors/dnd-move.in
  echo "48 26 22 ../48x48/dnd-none.png 0 ">> cursors/dnd-none.in
  echo "48 22 22 ../48x48/dotbox.png 0 ">> cursors/dotbox.in
  echo "48 22 22 ../48x48/fd_double_arrow.png 0 ">> cursors/fd_double_arrow.in
  echo "48 26 22 ../48x48/grabbing.png 0 ">> cursors/grabbing.in
  echo "48 18 10 ../48x48/hand2.png 0 ">> cursors/hand2.in
  echo "48 14 8 ../48x48/left_ptr.png 0 ">> cursors/left_ptr.in
  echo "48 14 26 ../48x48/left_side.png 0 ">> cursors/left_side.in
  echo "48 14 26 ../48x48/left_tee.png 0 ">> cursors/left_tee.in
  echo "48 8 10 ../48x48/link.png 0 ">> cursors/link.in
  echo "48 8 40 ../48x48/ll_angle.png 0 ">> cursors/ll_angle.in
  echo "48 40 40 ../48x48/lr_angle.png 0 ">> cursors/lr_angle.in
  echo "48 8 10 ../48x48/move.png 0 ">> cursors/move.in
  echo "48 14 46 ../48x48/pencil.png 0 ">> cursors/pencil.in
  echo "48 22 22 ../48x48/plus.png 0 ">> cursors/plus.in
  echo "48 24 42 ../48x48/question_arrow.png 0 ">> cursors/question_arrow.in
  echo "48 34 8 ../48x48/right_ptr.png 0 ">> cursors/right_ptr.in
  echo "48 40 26 ../48x48/right_side.png 0 ">> cursors/right_side.in
  echo "48 40 26 ../48x48/right_tee.png 0 ">> cursors/right_tee.in
  echo "48 22 36 ../48x48/sb_down_arrow.png 0 ">> cursors/sb_down_arrow.in
  echo "48 22 22 ../48x48/sb_h_double_arrow.png 0 ">> cursors/sb_h_double_arrow.in
  echo "48 14 26 ../48x48/sb_left_arrow.png 0 ">> cursors/sb_left_arrow.in
  echo "48 40 26 ../48x48/sb_right_arrow.png 0 ">> cursors/sb_right_arrow.in
  echo "48 24 8 ../48x48/sb_up_arrow.png 0 ">> cursors/sb_up_arrow.in
  echo "48 22 22 ../48x48/sb_v_double_arrow.png 0 ">> cursors/sb_v_double_arrow.in
  echo "48 22 22 ../48x48/tcross.png 0 ">> cursors/tcross.in
  echo "48 10 10 ../48x48/top_left_corner.png 0 ">> cursors/top_left_corner.in
  echo "48 36 10 ../48x48/top_right_corner.png 0 ">> cursors/top_right_corner.in
  echo "48 24 10 ../48x48/top_side.png 0 ">> cursors/top_side.in
  echo "48 24 10 ../48x48/top_tee.png 0 ">> cursors/top_tee.in
  echo "48 10 10 ../48x48/ul_angle.png 0 ">> cursors/ul_angle.in
  echo "48 36 10 ../48x48/ur_angle.png 0 ">> cursors/ur_angle.in
  echo "48 22 22 ../48x48/X_cursor.png 0 ">> cursors/X_cursor.in
  echo "48 22 22 ../48x48/xterm.png 0 ">> cursors/xterm.in
  echo "48 22 22 ../48x48/zoom-in.png 0 ">> cursors/zoom-in.in
  echo "48 22 22 ../48x48/zoom-out.png 0 ">> cursors/zoom-out.in

  echo "24 6 4 ../24x24/left_ptr_watch_1.png 50" > cursors/left_ptr_watch.in
  echo "24 6 4 ../24x24/left_ptr_watch_2.png 50" >> cursors/left_ptr_watch.in
  echo "24 6 4 ../24x24/left_ptr_watch_3.png 50" >> cursors/left_ptr_watch.in
  echo "24 6 4 ../24x24/left_ptr_watch_4.png 50" >> cursors/left_ptr_watch.in
  echo "24 6 4 ../24x24/left_ptr_watch_5.png 50" >> cursors/left_ptr_watch.in

  echo "32 8 5 ../32x32/left_ptr_watch_1.png 50" >> cursors/left_ptr_watch.in
  echo "32 8 5 ../32x32/left_ptr_watch_2.png 50" >> cursors/left_ptr_watch.in
  echo "32 8 5 ../32x32/left_ptr_watch_3.png 50" >> cursors/left_ptr_watch.in
  echo "32 8 5 ../32x32/left_ptr_watch_4.png 50" >> cursors/left_ptr_watch.in
  echo "32 8 5 ../32x32/left_ptr_watch_5.png 50" >> cursors/left_ptr_watch.in

  echo "48 12 8 ../48x48/left_ptr_watch_1.png 50" >> cursors/left_ptr_watch.in
  echo "48 12 8 ../48x48/left_ptr_watch_2.png 50" >> cursors/left_ptr_watch.in
  echo "48 12 8 ../48x48/left_ptr_watch_3.png 50" >> cursors/left_ptr_watch.in
  echo "48 12 8 ../48x48/left_ptr_watch_4.png 50" >> cursors/left_ptr_watch.in
  echo "48 12 8 ../48x48/left_ptr_watch_5.png 50" >> cursors/left_ptr_watch.in


  echo "24 12 12 ../24x24/watch_1.png 50" > cursors/watch.in
  echo "24 12 12 ../24x24/watch_2.png 50" >> cursors/watch.in
  echo "24 12 12 ../24x24/watch_3.png 50" >> cursors/watch.in
  echo "24 12 12 ../24x24/watch_4.png 50" >> cursors/watch.in
  echo "24 12 12 ../24x24/watch_5.png 50" >> cursors/watch.in
  echo "24 12 12 ../24x24/watch_6.png 50" >> cursors/watch.in
  echo "24 12 12 ../24x24/watch_7.png 50" >> cursors/watch.in
  echo "24 12 12 ../24x24/watch_8.png 50" >> cursors/watch.in
  echo "24 12 12 ../24x24/watch_9.png 50" >> cursors/watch.in
  echo "24 12 12 ../24x24/watch_10.png 50" >> cursors/watch.in

  echo "32 18 18 ../32x32/watch_1.png 50" >> cursors/watch.in
  echo "32 18 18 ../32x32/watch_2.png 50" >> cursors/watch.in
  echo "32 18 18 ../32x32/watch_3.png 50" >> cursors/watch.in
  echo "32 18 18 ../32x32/watch_4.png 50" >> cursors/watch.in
  echo "32 18 18 ../32x32/watch_5.png 50" >> cursors/watch.in
  echo "32 18 18 ../32x32/watch_6.png 50" >> cursors/watch.in
  echo "32 18 18 ../32x32/watch_7.png 50" >> cursors/watch.in
  echo "32 18 18 ../32x32/watch_8.png 50" >> cursors/watch.in
  echo "32 18 18 ../32x32/watch_9.png 50" >> cursors/watch.in
  echo "32 18 18 ../32x32/watch_10.png 50" >> cursors/watch.in

  echo "48 24 24 ../48x48/watch_1.png 50" >> cursors/watch.in
  echo "48 24 24 ../48x48/watch_2.png 50" >> cursors/watch.in
  echo "48 24 24 ../48x48/watch_3.png 50" >> cursors/watch.in
  echo "48 24 24 ../48x48/watch_4.png 50" >> cursors/watch.in
  echo "48 24 24 ../48x48/watch_5.png 50" >> cursors/watch.in
  echo "48 24 24 ../48x48/watch_6.png 50" >> cursors/watch.in
  echo "48 24 24 ../48x48/watch_7.png 50" >> cursors/watch.in
  echo "48 24 24 ../48x48/watch_8.png 50" >> cursors/watch.in
  echo "48 24 24 ../48x48/watch_9.png 50" >> cursors/watch.in
  echo "48 24 24 ../48x48/watch_10.png 50" >> cursors/watch.in

  cd cursors/
  xcursorgen bd_double_arrow.in bd_double_arrow
  xcursorgen bottom_left_corner.in bottom_left_corner
  xcursorgen bottom_right_corner.in bottom_right_corner
  xcursorgen bottom_side.in bottom_side
  xcursorgen bottom_tee.in bottom_tee
  xcursorgen circle.in circle
  xcursorgen color-picker.in color-picker
  xcursorgen copy.in copy
  xcursorgen cross.in cross
  xcursorgen crossed_circle.in crossed_circle
  xcursorgen crosshair.in crosshair
  xcursorgen dnd-ask.in dnd-ask
  xcursorgen dnd-copy.in dnd-copy
  xcursorgen dnd-link.in dnd-link
  xcursorgen dnd-move.in dnd-move
  xcursorgen dnd-none.in dnd-none
  xcursorgen dotbox.in dotbox
  xcursorgen fd_double_arrow.in fd_double_arrow
  xcursorgen grabbing.in grabbing
  xcursorgen hand2.in hand2
  xcursorgen left_ptr_watch.in left_ptr_watch
  xcursorgen left_ptr.in left_ptr
  xcursorgen left_side.in left_side
  xcursorgen left_tee.in left_tee
  xcursorgen link.in link
  xcursorgen ll_angle.in ll_angle
  xcursorgen lr_angle.in lr_angle
  xcursorgen move.in move
  xcursorgen pencil.in pencil
  xcursorgen plus.in plus
  xcursorgen question_arrow.in question_arrow
  xcursorgen right_ptr.in right_ptr
  xcursorgen right_side.in right_side
  xcursorgen right_tee.in right_tee
  xcursorgen sb_down_arrow.in sb_down_arrow
  xcursorgen sb_h_double_arrow.in sb_h_double_arrow
  xcursorgen sb_left_arrow.in sb_left_arrow
  xcursorgen sb_right_arrow.in sb_right_arrow
  xcursorgen sb_up_arrow.in sb_up_arrow
  xcursorgen sb_v_double_arrow.in sb_v_double_arrow
  xcursorgen tcross.in tcross
  xcursorgen top_left_corner.in top_left_corner
  xcursorgen top_right_corner.in top_right_corner
  xcursorgen top_side.in top_side
  xcursorgen top_tee.in top_tee
  xcursorgen ul_angle.in ul_angle
  xcursorgen ur_angle.in ur_angle
  xcursorgen watch.in watch
  xcursorgen X_cursor.in X_cursor
  xcursorgen xterm.in xterm
  xcursorgen zoom-in.in zoom-in
  xcursorgen zoom-out.in zoom-out
#存在软链接的鼠标图形
  ln -s hand2 e29285e634086352946a0e7090d73106
  ln -s fd_double_arrow fcf1c3c7cd4491d801f1e1c78f100000
  ln -s question_arrow d9ce0ab605698f320427677b458ad60b
  ln -s bd_double_arrow c7088f0f3e6c8088236ef8e1e3e70000
  ln -s hand2 9d800788f1b08800ae810202380a0822
  ln -s move 9081237383d90e509aa00f00170e968f
  ln -s link 640fb0e74195791501fd1ed57b41487f
  ln -s copy 6407b0e94181790501fd1e167b474872
  ln -s question_arrow 5c6cd98b3f3ebcb1f9c7f1c204630408
  ln -s move 4498f0e0c1937ffe01fd06f973665830
  ln -s left_ptr_watch 3ecb610c1bf2410f44200f48c40d3599
  ln -s link 3085a0e285430894940527032f8b26df
  ln -s sb_v_double_arrow 2870a09082c103050810ffdffffe0204
  ln -s sb_h_double_arrow 14fef782d02440884392942c11205230
  ln -s copy 1081e37283d90000800003c07f3ef6bf
  ln -s left_ptr_watch 08e8e1c95fe2fc01f976f1e063a24ccd
  ln -s crossed_circle 03b6e0fcb3499374a867c041f52298f0
  ln -s h_double_arrow 028006030e0e7ebffc7f7070c0600140
  ln -s v_double_arrow 00008160000006810000408080010102
  ln -s left_ptr arrow
  ln -s cross cross_reverse
  ln -s cross diamond_cross
  ln -s dotbox dot_box_mask
  ln -s sb_v_double_arrow double_arrow
  ln -s right_ptr draft_large
  ln -s right_ptr draft_small
  ln -s dotbox draped_box
  ln -s grabbing fleur
  ln -s hand2 hand
  ln -s hand2 hand1
  ln -s sb_h_double_arrow h_double_arrow
  ln -s question_arrow help
  ln -s dotbox icon
  ln -s question_arrow left_ptr_help
  ln -s X_cursor pirate
  ln -s dotbox target
  ln -s left_ptr top_left_arrow
  ln -s sb_v_double_arrow v_double_arrow

#
  rm *.in
  cd ..
